src/main/java/com/coveros/selenified/utilities/CombinedPDFReport.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.pdfbox.pdmodel.common.PDRectangle;
+import org.apache.pdfbox.pdmodel.font.PDFont;
+import org.apache.pdfbox.pdmodel.font.PDType1Font;
+import org.apache.pdfbox.pdmodel.interactive.action.PDActionGoTo;
+import org.apache.pdfbox.pdmodel.interactive.annotation.PDAnnotation;
+import org.apache.pdfbox.pdmodel.interactive.annotation.PDAnnotationLink;
+import org.apache.pdfbox.pdmodel.interactive.annotation.PDBorderStyleDictionary;
+import org.apache.pdfbox.pdmodel.interactive.documentnavigation.destination.PDPageDestination;
+import org.apache.pdfbox.pdmodel.interactive.documentnavigation.destination.PDPageFitWidthDestination;
+import org.testng.IReporter;
+import org.testng.ISuite;
+import org.testng.xml.XmlSuite;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.regex.Pattern;
+
+public class CombinedPDFReport implements IReporter {
+
+    static final float INCH = 72;
+
+    @Override
+    public void generateReport(List<XmlSuite> xmlSuites, List<ISuite> suites, String outputDirectory) {
+        if (Property.generatePDF()) {
+            try {
+                File folder = new File(outputDirectory + "/Surefire suite");
+
+                PDFMergerUtility pdfMerger = new PDFMergerUtility();
+                pdfMerger.setDestinationFileName(outputDirectory + "/Surefire suite/merged.pdf");
+                File[] testReports = listFilesMatching(folder, ".*\\.pdf");
+                for (File file : testReports) {
+                    pdfMerger.addSource(file);
+                }
+                pdfMerger.mergeDocuments(MemoryUsageSetting.setupMainMemoryOnly());
+                PDDocument document = PDDocument.load(new File(outputDirectory + "/Surefire suite/merged.pdf"));
+
+                PDPage tableOfContents = new PDPage(new PDRectangle(1000, testReports.length * 17));
+
+                float ph = tableOfContents.getMediaBox().getUpperRightY();
+
+                PDFont font = PDType1Font.HELVETICA_BOLD;
+                int fontSize = 12;
+                PDPageContentStream contents = new PDPageContentStream(document, tableOfContents);
+                contents.beginText();
+                contents.setFont(font, fontSize);
+                contents.newLineAtOffset(INCH, ph - INCH - fontSize);
+                for (File file : testReports) {
+                    contents.showText(file.getName());
+                    contents.newLineAtOffset(0, -15);
+                }
+                contents.endText();
+                contents.close();
+
+                // Now add the link annotation, so the click on "Jump to page three" works
+                for (int reportIndex = 1; reportIndex <= testReports.length; reportIndex++) {
+                    setLink(document, ph, tableOfContents, font, fontSize, reportIndex-1, "Go to " + testReports[reportIndex-1], reportIndex);
+                }
+
+                PDDocument newDoc = new PDDocument();
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Use try-with-resources or close this "PDDocument" in a "finally" clause. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/sonar/coding_rules#rule_key=squid%3AS2095)